MetaMask/metamask-extension
# split unit tests

## Explanation
@FrederikBolding recently enabled parallelism in our E2E test suites, this pr does the same for our unit tests. The changes are enumerated below for informational purposes and to aid review. 

### CircleCI Config changes
The original 'test-unit' job did a handful of things:
1. Ran mocha tests
2. Ran jest tests on the development folder
3. Ran jest tests across the rest of the application
4. Validated jest unit test coverage
5. Ran jest-it-up and failed if coverage had increased by a margin of 5

To simplify the process of test splitting I first split this job into a handful of others:
1. test-unit-mocha now runs mocha tests and nothing else, this job will run concurrently with the test-unit-jest job and not impact the overall CI time
2. test-unit-jest-development now runs jest tests on the development folder using the development jest config.
3. test-unit-jest-main now runs unit tests using the main jest config and uses jest's shard feature to split tests across N processes (specified in circle config as parallelism -- currently 12). Note that the command overrides coverage reporters and targets (more on this later).
4. test-unit-jest-validate-coverage now combines the coverage reports from the processes above and validates all of the coverage from the jest.config.js file targets.
7. jest-it-up now runs jest-it-up because combining coverage before running it is necessary.

### merge-coverage
jest.config.js has a 'coverageThresholds' setting which previously was sufficient for validating we hit our goals. Unfortuantely with that setting each individual test run when split across multiple processes fails because none of them reach the global coverage targets. To combat this, the script run in the new test file override our jest.config.js settings for their own purposes. 

In merge-coverage, we use istanbul's libraries to combine coverage reports and create new ones, which gets us to a final coverage report state. It then uses our jest.config.js's coverageThresholds to validate we have hit our targets. You can review the merge-coverage.js file for specifics but it uses fast-glob to find files that match the glob specified for the target of the coverage threshold. Note that for some reason permissions/flask coverage is not reported even without my change and so I explicitly ignored it for purposes of coverage targets. 

## Screenshots/Screencaps



### Before

![image](https://user-images.githubusercontent.com/4448075/201428210-12ea685b-bc99-43d0-afd5-28a22630924a.png)
![image](https://user-images.githubusercontent.com/4448075/201428249-ead4730a-5baa-4ade-9dc3-a6bd0ed7e7c0.png)

### After

![image](https://user-images.githubusercontent.com/4448075/201428103-8e824aa4-e0d7-483d-84f3-72bb8fe9e1d0.png)
![image](https://user-images.githubusercontent.com/4448075/201428144-fb587147-291f-4183-a229-d3486697cacf.png)


## Manual Testing Steps
1. You can visit the workflow for this branch and see all the runs and validate the success.


## Pre-merge author checklist
